[Luo Dan] iP

luodan01 opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/luodan01/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method:

```
public class Main {
    public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }
}
```

Commit:
Add Gradle support

luodan01 committed:
Add Greet, Echo, Exit

luodan01 committed:
Add List and Add functionalities

luodan01 committed:
Add Mark as Done

luodan01 committed:
Add ToDos, Events, Deadlines

luodan01 committed:
Add Automated Text UI Testing

luodan01 committed:
Add simple error handling

luodan01 committed:
Add support for deleting tasks

luodan01 committed:
Add more error handling

luodan01 committed:
Extract TaskList and Ui classes

luodan01 committed:
Add functionality to save tasks in hard disk automatically

luodan01 committed:
Add functionality for Duke to comprehend date and time

luodan01 committed:
Merge branch 'branch-Level-7' into branch-Level-8

luodan01 committed:
Merge branch 'branch-level-7'

luodan01 committed:
Merge changes

luodan01 committed:
Extract Command classes

luodan01 committed:
Divide classes into packages

luodan01 committed:
Divide classes into packages

luodan01 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	data/duke.txt
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/duke/Storage.java

luodan01 committed:
Remove tasks.txt

luodan01 committed:
Remove TaskType enum

NekrozQliphort commented on data/duke.txt:
```diff
@@ -0,0 +1,3 @@
+E | 0 | anEvent | 2100-09-09 12:23
+E | 1 | dedline | 2105-04-20
+E | 0 | anotherEvent | 3241-01-02 00:03
```
Should this file perhaps be placed in `.gitignore`?

NekrozQliphort commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,26 @@
+            throw new DukeException("☹ OOPS!!! That task doesn't exist.");
+        }
+        System.out.println(Ui.format("Noted. I've removed this task: \n\t" + tasks.get(idx) +
+                "\nNow you have " + ui.formatNumTasks(tasks.size() - 1) + " in the list."));
```
For the line wrapping, perhaps move the `+` to the next line?

luodan01 replied:
Fixed! 🙂

NekrozQliphort commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,67 @@
+        String taskInfo = params[1];
+        switch (type) {
+            case Event:
+                String[] eventInfo = taskInfo.split(" /at ");
```
Should the indentation for `case` be removed instead?

luodan01 replied:
Fixed! 🙂

NekrozQliphort commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    public String toString() {
+        return super.toString() + " (by: " +
+                date.format(DateTimeFormatter.ofPattern("MMM dd yyyy"))+ ")";
```
Perhaps it would be better if `String.format` was used here?

luodan01 replied:
Fixed! 🙂

NekrozQliphort commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,35 @@
+            return new DeleteCommand(Integer.parseInt(params[1]) - 1);
+        case ("todo"):
+        case ("event"):
+        case ("deadline"):
```
Should the `// Fallthrough` comment be added here?

luodan01 replied:
Fixed! 🙂

NekrozQliphort commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,32 @@
+        this.tasks = (existingTasks == null ? new ArrayList<>() : existingTasks);
+    }
+
+    public int size() {
```
Perhaps `size` could be named as `getSize` instead?

luodan01 replied:
Fixed! 🙂

NekrozQliphort commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,67 @@
+        return tasktypeInput.equals("event") ? TaskType.Event
+                : tasktypeInput.equals("deadline") ? TaskType.Deadline
+                : tasktypeInput.equals("todo") ? TaskType.ToDo
+                : TaskType.Invalid;
```
Would an `if else` statement help with clarity instead?

luodan01 replied:
Fixed by removing the function! 🙂

NekrozQliphort reviewed (commented):
Overall, I do like the abstractions being done here (especially `Command`), but I do think some of the string concatenation may be better done with `String.format`. Do remember to add Javadocs for the public methods as well. Other than that, the code looks good!

luodan01 committed:
Add tests

luodan01 committed:
Package the app as an executable JAR file

luodan01 committed:
Add Find functionality

luodan01 committed:
Add JavaDoc comments

luodan01 committed:
Tweak the code to comply with coding standard

luodan01 committed:
Tweak the code to comply with coding standard

luodan01 committed:
Tweak the code to comply with coding standard

idgrr commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,69 @@
+    protected String description;
+    protected boolean isDone;
+
+    public abstract String getTypeIndicator();
```
Since all task has a type, would it be better to have a type variable in the abstract class?

idgrr replied:
Not sure why my other comments didn't show up. Other things I commented on were 
1. Some missing header comments
2. Since all task has an isExit, would it be better to include it as a variable in the abstract class?
3. Very clear and readable method names. Nice!

luodan01 replied:
Thanks for your input! Regarding (2), I believe you are referring to the command classes. Unfortunately I'm unable to implement the suggestion since I made ```Command``` an interface

idgrr reviewed (commented):
Overall nicely done! Although some classes were missing javadocs, code was readable and easy to understand. If you disagree on certain points, do reply to them. I would be happy to discuss with you.

luodan01 committed:
Fix issues raised by Peer Review

luodan01 committed:
Merge branch 'add-gradle-support'

luodan01 committed:
Resolve merge conflicts

luodan01 committed:
Fix CheckStyle errors

luodan01 committed:
Add a GUI to Duke

luodan01 committed:
Add welcome message

luodan01 committed:
Merge branch 'branch-Level-10'

luodan01 committed:
Change Duke and User image

luodan01 committed:
Customise GUI

Duke chat bot does not have a customised icon and title.

 Having these customisations is more user-friendly since it would be easier to identify Duke from its taskbar icon and add personality to Duke.

Let's,
* add an icon
* add the title 'Duke'

luodan01 committed:
Add assertions

luodan01 committed:
Merge pull request #1 from luodan01/branch-A-Assertions

Add assertions

luodan01 committed:
Improve code quality

luodan01 committed:
Merge branch 'master' into branch-A-CodeQuality

luodan01 committed:
Merge pull request #2 from luodan01/branch-A-CodeQuality

Improve code quality

luodan01 committed:
Fix bye image bug

luodan01 committed:
Add functionality to archive tasks

luodan01 committed:
Merge pull request #4 from luodan01/branch-C-Archive

Branch c archive

luodan01 committed:
Improve GUI

luodan01 committed:
Finalise GUI

luodan01 committed:
Fix error message for adding tasks

luodan01 committed:
Add User guide

luodan01 committed:
Fix User guide

luodan01 committed:
Add Ui.png

luodan01 committed:
Fix line breaks

luodan01 committed:
Update User guide

luodan01 committed:
Merge remote-tracking branch 'origin/master'

luodan01 committed:
Fix typo in user guide

luodan01 committed:
Set theme jekyll-theme-slate
